[exporter/prometheusremotewriteexporter] Fix a go panic

falken opened the pull request:
Description: Protect against a Go Panic when the bucket count collection is empty (observed with histograms coming from Micrometer 1.9.1

Fixes #12168 #12168

codeboten reviewed (commented):
Thanks for the fix! Please add a changelog entry https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#changelog

falken commented:
> Thanks for the fix! Please add a changelog entry https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#changelog

Added, hopefully correctly

codeboten reviewed (commented):
Changelog looks good, please take a look at the lint failure:

```
Error: helper.go:345:9: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
make[2]: *** [../../../Makefile.Common:112: lint] Error 1
make[1]: *** [Makefile:184: pkg/translator/prometheusremotewrite] Error 2
make[1]: *** Waiting for unfinished jobs....
	} else {
```

MovieStoreGuy commented on pkg/translator/prometheusremotewrite/helper.go:
```diff
@@ -343,8 +343,10 @@ func addSingleHistogramDataPoint(pt pmetric.HistogramDataPoint, resource pcommon
 	} else {
-		cumulativeCount += pt.BucketCounts().At(pt.BucketCounts().Len() - 1)
-		infBucket.Value = float64(cumulativeCount)
+		if pt.BucketCounts().Len() > 0 {
```
My brief understanding of this code block is that this value ways always expected, I believe this check should be part of the original if statement as an or, so that there is always a value reported here.

falken replied:
I hadn't looked beyond fixing that error, I see now that it should just use the prior if statement to use the already specified default. Thank you for the catch!

MovieStoreGuy reviewed (changes requested):
This code path always intended to report a value, please make sure that a sensible default is report and add in a test to help ensure this change can't happen again.

falken committed:
[processor/prometheusremotewrite] Fix a go panic

Protect against a Go Panic when the bucket count collection is empty (observed with histograms coming from Micrometer 1.9.1

falken commented:
Okay, @MovieStoreGuy I have added a test, and verified that the fix works, however I can't figure out what's wrong with the assertion.

I get:

    helper_test.go:668:
                Error Trace:    /mnt/c/bench/dev/falken/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite/helper_test.go:668
                Error:          Not equal:
                                expected: map[string]*prompb.TimeSeries{"Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf":(*prompb.TimeSeries)(0xc00028ea80), "Histogram-__name__-foo_valid_Histogram_count":(*prompb.TimeSeries)(0xc00028eaf0), "Histogram-__name__-foo_valid_Histogram_sum":(*prompb.TimeSeries)(0xc00028eb60)}
                                actual  : map[string]*prompb.TimeSeries{"Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf":(*prompb.TimeSeries)(0xc00028ecb0), "Histogram-__name__-foo_valid_Histogram_count":(*prompb.TimeSeries)(0xc00028ec40), "Histogram-__name__-foo_valid_Histogram_sum":(*prompb.TimeSeries)(0xc00028ebd0)}

                                Diff:
                Test:           Test_addSingleHistogramDataPoint/histogram_without_buckets
--- FAIL: Test_addSingleHistogramDataPoint (0.00s)
    --- FAIL: Test_addSingleHistogramDataPoint/histogram_without_buckets (0.00s)

MovieStoreGuy commented:
I think you want to use `assert.EqualValues` instead of `assert.Exactly` since I suspect it is trying to enforce the same pointer address.

falken commented:
> I think you want to use `assert.EqualValues` instead of `assert.Exactly` since I suspect it is trying to enforce the same pointer address.

Both Exactly and EqualValues eventually use reflect.DeepEquals, which apparently returns false for the two maps.

falken commented:
@MovieStoreGuy Found the problem. Because the +Inf bucket has a value of NaN I can't compare them because NaN == NaN is false.

Aneurysm9 commented on pkg/translator/prometheusremotewrite/helper.go:
```diff
@@ -340,12 +340,13 @@ func addSingleHistogramDataPoint(pt pmetric.HistogramDataPoint, resource pcommon
 		Timestamp: time,
 	}
-	if pt.Flags().HasFlag(pmetric.MetricDataPointFlagNoRecordedValue) {
+	if pt.Flags().HasFlag(pmetric.MetricDataPointFlagNoRecordedValue) || pt.BucketCounts().Len() < 1 {
```
```suggestion
	if pt.Flags().HasFlag(pmetric.MetricDataPointFlagNoRecordedValue) || pt.BucketCounts().Len() == 0 {
```

The length of a slice can't be less than zero, so this is more representative of intent.  This makes clearer that this is a case of bad input that should probably be validated earlier and never make it this far.  The metric proto requires that there be [at least one entry](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L424-L426) in the bucket count list:

>   // The number of elements in bucket_counts array must be by one greater than
>  // the number of elements in explicit_bounds array.
>  repeated fixed64 bucket_counts = 6;

Even if the explicit bounds are empty then there must still be an entry in the `+Inf` bucket.

MovieStoreGuy replied:
In this scenario where something prior is failing to adhere to this, what would be the preferred scenario?

- In the event that a histogram metric has no values, then append an inf+ bucket with the value zero?
- In the event that a histogram metric has no values, ignore the metric? 

I think in terms of ensuring that data is correct here that this approach is the safest ensuring that data is correctly processed but I am not sure on the implications.

falken replied:
I update to == 0 and fixed the check to check for value.IsStaleNaN. I am not sure what the implication would be, but I would like to get the Go Panic fixed first.

Aneurysm9 replied:
I think what this PR is doing is probably wrong.  It is conflating the Prometheus use of a signaling `NaN` to convey staleness information with invalid histogram data points.  I suspect that the data point here should simply be ignored as early as possible.

falken replied:
The purpose of this PR is to stop the entire collector from crashing because of bad input. I made the PR as simple as possible in order to not affect any cases except a panic. This basically follows the same pattern as when the metric indicates it has no data "pmetric.MetricDataPointFlagNoRecordedValue". I understand that this data should probably be thrown out earlier, however that introduced possible additional behavior changes. With the lack of test coverage I felt this would have the least risk of introducing more problems because it will only affect data which would have resulted in a crash. 

If this is not an appropriate step forward I can close the PR and delete my fork.

MovieStoreGuy replied:
> If this is not an appropriate step forward I can close the PR and delete my fork.

You'll need to forgive us for being overly critical, the collector is depended on by a lot of users now so ensuring correctness of implementation is important. Don't let this be a discouragement to contribute but rather we really want you fix to work for all possible use cases / deployments. <3 

@Aneurysm9 , would you feel more comfortable with this change going in as is if there was a follow up issue to address the actual cause?

Aneurysm9 commented on pkg/translator/prometheusremotewrite/helper_test.go:
```diff
@@ -593,3 +593,84 @@ func TestMostRecentTimestampInMetric(t *testing.T) {
+			assert.Exactly(t, tt.expected["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Labels, tsMap["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Labels)
+			assert.Equal(t, len(tt.expected["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Samples), 1)
+			assert.Exactly(t, tt.expected["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Samples[0].Timestamp, tsMap["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Samples[0].Timestamp)
+			assert.True(t, math.IsNaN(tsMap["Histogram-__name__-foo_valid_Histogram_bucket-le-+Inf"].Samples[0].Value))
```
This is not expected to be any `NaN`, but a particular signaling `NaN`.  Use [`value.IsStaleNan()`](https://pkg.go.dev/github.com/prometheus/prometheus/pkg/value#IsStaleNaN) to validate.

MovieStoreGuy commented:
> @MovieStoreGuy Found the problem. Because the +Inf bucket has a value of NaN I can't compare them because NaN == NaN is false. I made the test explicitly test this case which should work.

I am glad you saw that because I did not.

Thank you for taking the time to fix this.

falken commented:
Apparently this is moot @Aneurysm9 @MovieStoreGuy since PR #12854
